src/Microsoft.VisualStudio.ProjectSystem.Managed/Threading/SemaphoreSlimExtensions.cs

```diff
@@ -57,5 +57,19 @@ await joinableTask.Task
                 }
             }
         }
+
+        public static async Task ExecuteWithinLockAsync(this SemaphoreSlim semaphore, JoinableTaskCollection collection, JoinableTaskFactory factory, Action action)
+        {
+            // Join the caller to our collection, so that if the lock is already held by another task that needs UI 
+            // thread access we don't deadlock if we're also being waited on by the UI thread. For example, when CPS
+            // is draining critical tasks and is waiting us.
```

Comment:
do you mean `is awaiting us`?

Reply:
Not sure - copied from the `Task<T>` version.  @davkean should this be "is awaiting us", or "is waiting for us"?